File: 1352-product-of-the-last-k-numbers/main.py
class ProductOfNumbers:

    def __init__(self):
        self.prefix_product = [1]
        self.size = 0

    def add(self, num: int) -> None:
        if num == 0:
            self.prefix_product = [1]
            self.size = 0
        else:
            self.prefix_product.append(self.prefix_product[self.size] * num)
            self.size += 1

    def getProduct(self, k: int) -> int:
        if k > self.size:
            return 0
        
        return (
            self.prefix_product[self.size] // self.prefix_product[self.size - k]
        )

File: 1352-product-of-the-last-k-numbers/main.cpp
class ProductOfNumbers {
public:
    vector<int> prefix_product;
    int size;

    ProductOfNumbers() {
        prefix_product.push_back(1);
        size = 0;
    }
    
    void add(int num) {
        if (num == 0) {
            prefix_product.clear();
            prefix_product.push_back(1);
            size = 0;
        }else {
            prefix_product.push_back(prefix_product.back() * num);
            size++;
        }
    }
    
    int getProduct(int k) {
        if (k > size) {
            return 0;
        }

        return prefix_product.back() / prefix_product[prefix_product.size() - 1 - k];
    }
};

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers* obj = new ProductOfNumbers();
 * obj->add(num);
 * int param_2 = obj->getProduct(k);
 */
